File: ios/ReactNativeCharts/DataExtract.swift
//  Created by xudong wu on 02/03/2017.
//  Copyright © 2017 wuxudong. All rights reserved.
//

import Foundation
import Charts
import SwiftyJSON

public class DataExtract {
    func extract(_ data: JSON) -> ChartData?  {
        if data["dataSets"].array == nil {
            return nil;
        }
        
        let chartData = createData();
        
        
        let dataSets = data["dataSets"].arrayValue;
        
        for (_, dataSet) in dataSets.enumerated() {
            
            let values = dataSet["values"].arrayValue;
            let label = dataSet["label"].stringValue;
            
            let entries = createEntries(values: values);
            
            let chartDataSet = createDataSet(entries: entries, label: label);
            
            if dataSet["config"].dictionary != nil {
                dataSetConfig(dataSet: chartDataSet, config: dataSet["config"])
            }
            
            chartData.addDataSet(chartDataSet);
        }
        
        if data["config"].dictionary != nil {
            dataConfig(data: chartData, config: data["config"])
        }

        
        return chartData

    }
    
    func createEntries(values: [JSON]) -> [ChartDataEntry] {
        var entries = [ChartDataEntry]();
        
        for (index, value) in values.enumerated() {
            if value.null == nil {
                entries.append(createEntry(values: values, index: index))
            }
        }
        
        return entries;
        
    }

    
    func createData() -> ChartData {
        fatalError("subclass should override this function")
    }
    
    func dataConfig(data: ChartData, config: JSON) {}
    
    func createDataSet(entries: [ChartDataEntry]?, label: String?) -> IChartDataSet {
        fatalError("subclass should override this function")
    }
    
    func dataSetConfig(dataSet: IChartDataSet, config: JSON) {
        fatalError("subclass should override this function")
    }
    
    func createEntry(values: [JSON], index: Int) -> ChartDataEntry {
        fatalError("subclass should override this function")
    }

    
}

File: android/src/main/java/com/github/wuxudong/rncharts/data/DataExtract.java
package com.github.wuxudong.rncharts.data;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableType;
import com.github.mikephil.charting.charts.Chart;
import com.github.mikephil.charting.data.ChartData;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.interfaces.datasets.IDataSet;
import com.github.wuxudong.rncharts.utils.BridgeUtils;

import java.util.ArrayList;

/**
 * Created by xudong on 02/03/2017.
 */

public abstract class DataExtract<D extends ChartData, U extends Entry> {

    public D extract(Chart chart, ReadableMap propMap) {
        if (!BridgeUtils.validate(propMap, ReadableType.Array, "dataSets")) {
            return null;
        }

        D chartData = createData();


        ReadableArray dataSets = propMap.getArray("dataSets");
        for (int i = 0; i < dataSets.size(); i++) {
            ReadableMap dataSetReadableMap = dataSets.getMap(i);

            // TODO validation
            ReadableArray values = dataSetReadableMap.getArray("values");
            String label = dataSetReadableMap.getString("label");

            ArrayList<U> entries = createEntries(values);

            IDataSet<U> dataSet = createDataSet(entries, label);

            if (BridgeUtils.validate(dataSetReadableMap, ReadableType.Map, "config")) {
                dataSetConfig(chart, dataSet, dataSetReadableMap.getMap("config"));
            }

            chartData.addDataSet(dataSet);
        }

        if (BridgeUtils.validate(propMap, ReadableType.Map, "config")) {
            dataConfig(chartData, propMap.getMap("config"));
        }


        return chartData;
    }

    abstract D createData();

    void dataConfig(D data, ReadableMap config) {
    }

    abstract IDataSet<U> createDataSet(ArrayList<U> entries, String label);

    abstract void dataSetConfig(Chart chart, IDataSet<U> dataSet, ReadableMap config);

    ArrayList<U> createEntries(ReadableArray yValues) {
        ArrayList<U> entries = new ArrayList<>(yValues.size());
        for (int j = 0; j < yValues.size(); j++) {
            if (!yValues.isNull(j)) {
                entries.add(createEntry(yValues, j));
            }
        }
        return entries;
    }

    abstract U createEntry(ReadableArray values, int index);


}
